### PR TITLE
[chore] Optimize batch feature saving

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

Optimize batch feature saving by
* regularly clearing up the global query & operation structure cache
* cache `sorted_node_names` at `QueryGraphModel` level

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/worker/util/batch_feature_creator.py [hunk 1]
```diff
@@ -351,9 +362,20 @@ async def batch_feature_create(
         )
         await ranged_progress_update(0, "Started saving features")
 
-        output_feature_ids = []
+        output_feature_ids: List[PydanticObjectId] = []
         inconsistent_feature_names = []
         for i, feature_item in enumerate(payload.features):
+            if (len(output_feature_ids) + 1) % graph_clear_frequency == 0:
```
Comment (YS-L) [inline featurebyte/worker/util/batch_feature_creator.py:368]: Does `output_feature_ids` include features that were already saved and skipped?

Comment (gee-senbong) [reply featurebyte/worker/util/batch_feature_creator.py:368]: Good catch. Could use another counter to account for the non-skipped feature. Will update this part.

File: featurebyte/common/utils.py [hunk 2]
```diff
@@ -471,3 +474,29 @@ def is_server_mode() -> bool:
...
+    Generator[None, None, None]
+        A context manager that logs the execution time with the given message and logger.
+    """
+    start_time = time.time()
+    try:
+        yield
+    finally:
+        end_time = time.time()
+        duration = end_time - start_time
+        logger.info(f"{message}: {duration} seconds")
```
Comment (YS-L) [inline featurebyte/common/utils.py:502]: Might be useful if this utility allows specifying extra dict that will be passed to logger.

**Review by YS-L — COMMENTED** [review]
Curious if there is a timing comparison before vs after the changes?

File: featurebyte/query_graph/model/graph.py [hunk 3]
```diff
@@ -115,6 +116,19 @@ def sorted_node_names_by_ref(self) -> List[str]:
             self._update_cache()
         return self._sorted_node_names_by_ref
 
+    @property
+    def sorted_node_names(self) -> List[str]:
```
Comment (YS-L) [inline featurebyte/query_graph/model/graph.py:120]: Was this also a critical path that slowed down the graph operations?

**Review by YS-L — APPROVED** [review]

File: featurebyte/query_graph/model/graph.py (hunk 3 above)
Comment (gee-senbong) [reply featurebyte/query_graph/model/graph.py:120]: Yes, when performing the quick prune operation. This is called multiple times.

**Comment by gee-senbong** [comment]
> Curious if there is a timing comparison before vs after the changes?

Previously, saving 3841 features takes 1 h 22 minutes -> 1.281s per feature
After clearing global graph (without sorted node name caching), saving 4941 features takes 1h 21 minutes -> 0.984s per feature
Current version (still saving), saving 2472 takes 38 minutes -> 0.9223 per feature

**Pull request merged**